### PR TITLE
feat: add signed division

### LINKED ISSUE #1790: Implement signed division
### Problem

Currently the acir_gen pass does not have signed division. It is now a todo in the compiler.

### Happy Case

Implement signed division

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

Implement signed division by convertir the inputs to unsigned integers, doing the euclidian division there and transforming back the result for the signed numbers

## Problem\*

Resolves #1790

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs [hunk 1]
```diff
@@ -431,6 +433,37 @@ impl AcirContext {
...
     }
 
+    /// Returns the quotient and remainder such that lhs = rhs * quotient + remainder
+    /// and |remainder| < |rhs|
+    /// and remainder has the same sign than lhs
+    /// Note that this is not the euclidian division, where we have instead remainder < |rhs|
+    ///
+    ///
+    ///
+    ///
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs:443]: Why does signed division return two values?

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs [hunk 2]
```diff
@@ -307,6 +307,56 @@ impl GeneratedAcir {
         Ok(limb_witnesses)
     }
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs:309]: I finalised the algorithm and explained what it does.

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs:443]: it returns the quotient and the remainder, like the unsigned division

File: crates/nargo_cli/tests/test_data_ssa_refactor/unsigned_division/src/main.nr [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// -7/3  = -2
+// -7/-3 = 2
+//  7/-3 = -2
+fn main(mut x: i32, mut y: i32, mut z: i32) {
+    // 7/3 = 2
+    assert(x / y == z);
+
+    // -7/3 = -2
+    x = 0-x;
+    z = 0-z;
```
Comment (kevaundray) [inline crates/nargo_cli/tests/test_data_ssa_refactor/unsigned_division/src/main.nr:14]: can we not re-use x, y, z. minus_x, minus_y etc would be clearer

**Review by kevaundray — COMMENTED** [review]
Minor comments, generally looks okay

File: crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs [hunk 4]
```diff
@@ -431,6 +433,37 @@ impl AcirContext {
...
+        bit_size: u32,
+    ) -> Result<(AcirVar, AcirVar), AcirGenError> {
+        let lhs_data = &self.vars[&lhs].clone();
+        let rhs_data = &self.vars[&rhs].clone();
+
+        let lhs_expr = lhs_data.to_expression();
+        let rhs_expr = rhs_data.to_expression();
+        let l_witness = self.acir_ir.get_or_create_witness(&lhs_expr);
+        let r_witness = self.acir_ir.get_or_create_witness(&rhs_expr);
+        assert_ne!(bit_size, 0);
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs]: Done

File: crates/nargo_cli/tests/test_data_ssa_refactor/unsigned_division/src/main.nr (hunk 3 above)
Comment (guipublic) [reply crates/nargo_cli/tests/test_data_ssa_refactor/unsigned_division/src/main.nr]: Done

**Review by kevaundray — APPROVED** [review]

**Pull request merged**